cases/serializers.py

```
@@ -540,9 +652,34 @@ class Meta:
+        afterwards — so a validation failure never leaves an orphaned S3 object.
+        """
+        from cases.services.source_files import store_file_as_link
+
+        uploaded_file = validated_data.pop("uploaded_file", None)
+        upload_role = validated_data.pop("upload_role", SourceLinkRole.RAW.value)
+
+        instance = super().create(validated_data)
+
+        if uploaded_file is not None:
+            link = store_file_as_link(uploaded_file, role=upload_role)
+            instance.url = list(instance.url or []) + [link]
```

Ruff (0.15.15)

[warning] Consider `[*list(instance.url or []), link]` instead of concatenation

Replace with `[*list(instance.url or []), link]`

(RUF005)